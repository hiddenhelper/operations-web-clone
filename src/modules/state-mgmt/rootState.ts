import { History } from 'history';

import { IState as IAuthState } from './auth/state';
import { IState as IGeneralState } from './general/state';
import { IState as IClientState } from './client/state';
import { IState as IProjectState } from './project/state';
import { IState as IUserState } from './user/state';
import { IState as IAccessControlSystemState } from './access-control-system/state';
import { IState as IBadgePrintingSystemState } from './badge-printer-system/state';
import { IState as IWorkerState } from './worker/state';
import { IState as ICertificationState } from './certification/state';
import { IState as IFileState } from './file/state';
import { IState as IBadgeState } from './badge/state';
import { IState as IStatisticsState } from './statistics/state';
import { IState as ITrainingState } from './training/state';
import { IState as IPaymentState } from './payment/state';
import { IState as IInvoiceState } from './invoice/state';
import { ApiService } from '../services/ApiService';
import { AuthService } from '../services/AuthService';
import { PrinterService } from '../services/PrinterService';
import { FileService } from '../services/FileService';
import { Logger } from '../services/Logger';
import { IProcoreState } from './procore/state';
import { IState as IProjectNewState } from './project-new/state';

export interface IAction {
  type: string;
  payload: any;
}

export interface IRootState {
  auth: IAuthState;
  user: IUserState;
  client: IClientState;
  project: IProjectState;
  accessControlSystem: IAccessControlSystemState;
  badgePrinterSystem: IBadgePrintingSystemState;
  worker: IWorkerState;
  certification: ICertificationState;
  training: ITrainingState;
  invoice: IInvoiceState;
  file: IFileState;
  badge: IBadgeState;
  statistics: IStatisticsState;
  payment: IPaymentState;
  general: IGeneralState;
  router?: { location: { pathname: string; search: string; hard: string; key: string }; action: string };
  procore: IProcoreState;
  projectNew: IProjectNewState;
}

export interface IEpicDependencies {
  history: History;
  apiService: ApiService;
  authService: AuthService;
  printerService: PrinterService;
  fileService: FileService;
  logger: typeof Logger;
}
