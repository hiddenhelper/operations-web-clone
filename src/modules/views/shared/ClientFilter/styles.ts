import { createStyles, makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles(theme =>
  createStyles({
    autocompleteFilterPosition: {
      '& .MuiAutocomplete-root': {
        position: 'absolute',
      },

      '& .MuiAutocomplete-popper': {
        position: 'absolute',
        top: 65,
      },
    },
  })
);
