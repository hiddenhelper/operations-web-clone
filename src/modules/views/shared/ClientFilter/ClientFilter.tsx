import React, { memo, useCallback, useEffect, useMemo } from 'react';

import Box from '@material-ui/core/Box';

import AutocompleteFilter from '../AutocompleteFilter';

import { ClientModel, GeneralModel } from '../../../models';
import { getConditionalDefaultValue } from '../../../../utils/generalUtils';

import { tableGlobalStyles } from '../../../../assets/styles';
import { useStyles } from './styles';

export interface IClientFilterProps {
  queryParams: GeneralModel.IQueryParams;
  clientMap?: GeneralModel.IEntityMap<ClientModel.IClientProject>;
  projectId: string;
  isFcaUser: boolean;
  setQueryParams: (params: GeneralModel.IQueryParams) => void;
  fetchClientList?: (id: string, query: GeneralModel.IQueryParams) => void;
}

const ClientFilter = ({ queryParams, clientMap = {}, projectId, isFcaUser, setQueryParams, fetchClientList }: IClientFilterProps) => {
  const classes = useStyles();
  const tableGlobalClasses = tableGlobalStyles();

  const clientLabel = useMemo(() => getConditionalDefaultValue(isFcaUser, 'All Clients', 'All Companies'), [isFcaUser]);

  const entityTabList: GeneralModel.INamedEntity[] = useMemo(() => Object.values(clientMap), [clientMap]);
  const clientList = useMemo(() => entityTabList.reduce((tot, item) => [...tot, { id: item?.id, name: item?.name }], [{ id: '', name: clientLabel }]), [
    entityTabList,
    clientLabel,
  ]);
  const tabEntityMap = useMemo(() => entityTabList.reduce((tot, item) => ({ ...tot, [item?.id]: item?.name }), {}), [entityTabList]);

  const onFilterClientChange = useCallback(
    (companyId: string) => {
      setQueryParams({ ...queryParams, companyId, page: 1 });
    },
    [setQueryParams, queryParams]
  );

  useEffect(() => {
    fetchClientList(projectId, { ...queryParams, page: 1, limit: 30 });
  }, [fetchClientList, projectId]); // eslint-disable-line

  return (
    <Box className={`${tableGlobalClasses.filterStatusContainer} ${classes.autocompleteFilterPosition}`}>
      <AutocompleteFilter
        value={queryParams.companyId}
        label={getConditionalDefaultValue(queryParams.companyId, tabEntityMap[queryParams.companyId], clientLabel)}
        optionList={clientList}
        onChange={onFilterClientChange}
      />
    </Box>
  );
};

export default memo(ClientFilter);
