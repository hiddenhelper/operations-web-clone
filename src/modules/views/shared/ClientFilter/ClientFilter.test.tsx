import React from 'react';
import { MemoryRouter } from 'react-router';
import { act, fireEvent, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createMockStore } from 'redux-test-utils';
import ClientFilter, { IClientFilterProps } from './ClientFilter';
import { getClient_1, getProject_1 } from '../../../../test/entities';
import { getInitialState } from '../../../../test/rootState';

describe('ClientFilter', () => {
  let props: IClientFilterProps;

  beforeEach(() => {
    props = {
      queryParams: { page: 1, limit: 30 },
      clientMap: {
        [getClient_1().id]: getClient_1() as any,
      },
      projectId: getProject_1().id,
      isFcAdmin: true,
      setQueryParams: jest.fn(),
      fetchClientList: jest.fn(),
    };
  });

  it('should render', () => {
    const wrapper = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <ClientFilter {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(props.fetchClientList).toHaveBeenCalled();
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should render default', () => {
    props.clientMap = undefined;
    const wrapper = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <ClientFilter {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should change client filter', () => {
    const wrapper = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <ClientFilter {...props} />
        </MemoryRouter>
      </Provider>
    );

    act(() => {
      fireEvent.click(wrapper.getAllByText('All Clients')[0]);
    });

    act(() => {
      fireEvent.click(wrapper.getByText('Robert C. Martin'));
    });

    expect(props.setQueryParams).toHaveBeenCalled();
  });
});
