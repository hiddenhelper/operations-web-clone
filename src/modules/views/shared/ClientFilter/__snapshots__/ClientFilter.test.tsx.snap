// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientFilter should render 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-41 makeStyles-filterStatusContainer-16 makeStyles-autocompleteFilterPosition-1"
  >
    <div
      class="makeStyles-floatingFilterWrapper-22 makeStyles-filterWrapper-21 makeStyles-buttonContainer-43 floatingFilterWrapper"
    >
      <span
        class="makeStyles-floatingFilterButton-23"
        data-testid="controlled-button"
        tabindex="-1"
      >
        <button
          aria-controls="false"
          aria-haspopup="true"
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-buttonFilter-31"
          data-testid="button-filter-open"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            All Clients
            <span
              class="MuiButton-endIcon MuiButton-iconSizeMedium"
            >
              <svg
                class="makeStyles-dropdownIcon-25"
              >
                expand.svg
              </svg>
            </span>
          </span>
        </button>
      </span>
    </div>
  </div>
</div>
`;

exports[`ClientFilter should render default 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-87 makeStyles-filterStatusContainer-62 makeStyles-autocompleteFilterPosition-47"
  >
    <div
      class="makeStyles-floatingFilterWrapper-68 makeStyles-filterWrapper-67 makeStyles-buttonContainer-89 floatingFilterWrapper"
    >
      <span
        class="makeStyles-floatingFilterButton-69"
        data-testid="controlled-button"
        tabindex="-1"
      >
        <button
          aria-controls="false"
          aria-haspopup="true"
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-buttonFilter-77"
          data-testid="button-filter-open"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            All Clients
            <span
              class="MuiButton-endIcon MuiButton-iconSizeMedium"
            >
              <svg
                class="makeStyles-dropdownIcon-71"
              >
                expand.svg
              </svg>
            </span>
          </span>
        </button>
      </span>
    </div>
  </div>
</div>
`;
